nexus/src/app/background/tasks/instance_reincarnation.rs

```diff
@@ -0,0 +1,674 @@
+
+pub struct InstanceReincarnation {
+    datastore: Arc<DataStore>,
+    sagas: Arc<dyn StartSaga>,
+    /// The maximum number of concurrently executing instance-start sagas.
+    concurrency_limit: NonZeroU32,
+}
+
+const DEFAULT_MAX_CONCURRENT_REINCARNATIONS: NonZeroU32 =
+    match NonZeroU32::new(16) {
+        Some(n) => n,
+        None => unreachable!(), // 16 > 0
+    };
+
+impl BackgroundTask for InstanceReincarnation {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        Box::pin(async move {
+            let mut status = InstanceReincarnationStatus::default();
+            self.actually_activate(opctx, &mut status).await;
+            if !status.restart_errors.is_empty() || status.query_error.is_some()
+            {
+                error!(
+                    &opctx.log,
+                    "instance reincarnation completed with errors!";
+                    "instances_found" => status.instances_found,
+                    "instances_reincarnated" => status.instances_reincarnated.len(),
+                    "instances_changed_state" => status.changed_state.len(),
+                    "query_error" => ?status.query_error,
+                    "restart_errors" => status.restart_errors.len(),
+                );
+            } else {
+                info!(
+                    &opctx.log,
+                    "instance reincarnation completed";
+                    "instances_found" => status.instances_found,
+                    "instances_reincarnated" => status.instances_reincarnated.len(),
+                    "instances_changed_state" => status.changed_state.len(),
+                );
+            }
+            serde_json::json!(status)
+        })
+    }
+}
+
+impl InstanceReincarnation {
+    pub(crate) fn new(
+        datastore: Arc<DataStore>,
+        sagas: Arc<dyn StartSaga>,
+    ) -> Self {
+        Self {
+            datastore,
+            sagas,
+            concurrency_limit: DEFAULT_MAX_CONCURRENT_REINCARNATIONS,
+        }
+    }
+
+    async fn actually_activate(
+        &mut self,
+        opctx: &OpContext,
+        status: &mut InstanceReincarnationStatus,
+    ) {
+        let mut total_sagas_started = 0;
+        let mut running_sagas =
+            Vec::with_capacity(self.concurrency_limit.get() as usize);
+        let serialized_authn = authn::saga::Serialized::for_opctx(opctx);
+
+        loop {
+            let maybe_batch = self
+                .datastore
+                .find_reincarnatable_instances(
+                    opctx,
+                    self.concurrency_limit,
+                    // Any instances which we've already attempted to start and
+                    // couldn't should be excluded from the query, to avoid
+                    // looping forever when an instance is in a permanently
+                    // non-restartable state.
+                    status.restart_errors.keys().copied(),
```

Comment:
Ah, yeah, I think you're right that this could result in an *enormous* query string getting passed to the DB and it's probably better to just change the state of the db record when a restart attempt fails. I'll change that.

Reply:
Rereading the `instance-start` saga, it looks like the instance gets its `time_last_auto_restarted`  timestamp set in `sis_move_to_starting`, and that doesn't get cleared if the start saga unwinds. So, really, the question is what we ought to do if the start saga fails before `sis_move_to_starting` --- when we attempt to allocate a Propolis server and IP --- or if _preparing_ the saga DAG fails (which, AFAIK, *shouldn't* ever happen). 

IMO, it's probably correct to not enforce an additional hour-long cooldown when `sis_alloc_server` or `sis_alloc_propolis_ip` fails. That doesn't mean the instance was reincarnated and failed again, it just means that there weren't presently sufficient resources available for that instance, and we probably _should_ try again the next time the task is activated, in case additional resources have become available and we can now find a place to put it? But, we don't want to keep seeing it on *this* activation, because then we'll just keep looping forever. Hmm...

Reply:
Oh, additionally...I realized that we _won't_ currently attempt to reincarnate instances that have had an `instance-start` saga failed if they've advanced past `sis_move_to_starting`, because then they will no longer be in `InstanceState::Failed`, and will instead be in `InstanceState::NoVmm` with a pointer to a `VmmState::SagaUnwound` VMM. I think we might want to also make instances with `SagaUnwound` VMMs eligible for reincarnation at some point, but I don't know if I want to deal with that in _this_ branch.

Reply:
It's also occurred to me that we probably don't want to punish an instance with an hour-long cooldown when it's `SagaUnwound`, since IIRC, all failures which move an instance to `SagaUnwound` will have occurred *before* the instance is actually incarnated on a VMM, so it can't possibly be the guest's fault in that case? It should probably have to wait to be reincarnated, too, just to avoid hot-looping them, but probably not as long as instances that actually made it to incarnation and then died...

Reply:
I think for now I'm probably going to just change the `instance-start` saga to add an action that sets the last-auto-restart timestamp before attempting to do anything else, if the start saga was created by an automated restart. That way, instances no longer appear in the `find_reincarnatable_instances` query if, and only if, a start saga for that instance has been started. I think we'll eventually want to consider auto-restarting instances with `SagaUnwound` VMMs, but I'll save that for a subsequent change.

Reply:
@gjcolombo commit https://github.com/oxidecomputer/omicron/pull/6503/commits/d3aa320889ade3a38f2d8fc2d233bf62fbb6243e changes the `instance-start` saga to immediately set the auto-restart timestamp as soon as the saga has started, if it was triggered by auto-restart, making it no longer show up in the query. I *think* this is correct, although adding an additional instance runtime state mutation *before* the action that moves it to `Starting` feels mildly sketchy...what do you think?